### PR TITLE
Ignore EPIPE and EOF errors on child stdin stream

### PR DESCRIPTION
**Pull request by niik**
This was discussed in https://github.com/desktop/desktop/pull/4027#issuecomment-366213276.

If a child process closes stdin or terminates while a stream write operation is in progress the error from stdin closing will be thrown before the exit callback on the process. Without an error handler subscribed to the stream the default behavior of Node is to re-throw the error which then ends up in the global uncaught error handler and then we crash (and by we I mean desktop/desktop).

The context should be fairly well laid out in the PR itself.

### CODE REVIEW

File: lib/git-process.ts [hunk 1]
```diff
@@ -247,3 +251,54 @@ export class GitProcess {
...
+
+    // Nope, this is something else. Are there any other error listeners
+    // attached than us? If not we'll have to mimic the behavior of
+    // EventEmitter.
+    //
+    // See https://nodejs.org/api/errors.html#errors_error_propagation_and_interception
+    //
+    // "For all EventEmitter objects, if an 'error' event handler is not
+    //  provided, the error will be thrown"
+    if (process.stdin.listeners('error').length > 1) {
```
Comment (shiftkey) [inline lib/git-process.ts]: > Are there any other error listeners attached than us? If not we'll have to mimic the behavior of EventEmitter.

This mean we should be doing the inverse of this check, right?

```ts
if (process.stdin.listeners('error').length <= 1) {
   throw err
}
```

This feels similar to what `epipebomb` [does](https://github.com/mhart/epipebomb/blob/8c1e6d83a32678f77daf3dad617122291c984517/epipebomb.js#L10-L12):

```js
    // If there's more than one error handler (ie, us),
    // then the error won't be bubbled up anyway
    if (stream.listeners('error').length <= 1) {
      ...
    }
```

Comment (niik) [reply lib/git-process.ts]: Doh, right you are. Nice catch.

**Comment by niik** [comment]
The test [caught ECONNRESET on Linux](https://travis-ci.org/desktop/dugite/jobs/343712728). Not sure if it's just another edge case or if it's linux specific but I'm including it as an error we'll handle.

screen shot 2018-02-20 at 10 23 04

**Comment by niik** [comment]
The more I think about it though I wonder if `ECONNRESET` came from stdout/stderr?

**Comment by niik** [comment]
> The more I think about it though I wonder if ECONNRESET came from stdout/stderr?

Fascinating, according to that build linux build failure from where I suppressed stdout/stderr the error does stem from stdin.
